contracts/vested-escrow/VeERC20v2.sol

```diff
@@ -0,0 +1,208 @@
+        if (t < (tgeTimestamp + cliffSeconds)) {
+            // there's nothing additional vested yet
+        } else { // if after the cliff
+            // if it's beyond vesting time = everything is vested
+            if ((tgeTimestamp + cliffSeconds + vestingSeconds) < t) {
+                vested = N;
+            } else {
+                // how much second passed after cliff
+                uint256 timePassed = (t - (tgeTimestamp + cliffSeconds));
+                // last full vesting part (i.e 13 month => k = 12 month with vestingFrequencySeconds = 6 month)
+                uint256 fullVestingPart = timePassed / vestingFrequencySeconds * vestingFrequencySeconds;
+
+                // (N - vested) = tokens left to be vested after tge
+                vested = vested + (N - vested) * fullVestingPart / vestingSeconds;
+            }
+        }
+
+        // the answer is how much is vested in total minus how much already withdrawn
+        return vested <= alreadyWithdrawn[account] ? 0 : vested - alreadyWithdrawn[account];
+    }
+
+    function claim() external {
+        uint256 claimAmount = availableToClaim(msg.sender);
+        if (claimAmount == 0) revert ClaimZeroError();
+        alreadyWithdrawn[msg.sender] = alreadyWithdrawn[msg.sender] + claimAmount;
+        _burn(msg.sender, claimAmount);
+        _transferVotingUnits(msg.sender, address(0), claimAmount);
+        LOCKED_TOKEN.safeTransfer(msg.sender, claimAmount);
+
+        emit Claimed(msg.sender, claimAmount);
+    }
+
+    // TODO: check if we need that and it's secure
+    function donateTokens(address toDAO) external {
+        if (hasRole(DAO_ROLE, toDAO) == false) revert BadDAOAddress(toDAO);
+        uint256 balance = super.balanceOf(msg.sender);
+        if (balance == 0) revert NoBalance();
+        alreadyWithdrawn[msg.sender] = alreadyWithdrawn[msg.sender] + balance;
+        _burn(msg.sender, balance);
+        _transferVotingUnits(msg.sender, address(0), balance);
+        LOCKED_TOKEN.safeTransfer(toDAO, balance);
+    }
+
+    function rescue(ERC20Upgradeable token) external onlyRole(DAO_ROLE) {
+        // TODO: check if we need that
+        require(token != LOCKED_TOKEN, "No allowed to rescue this token");
+        // allow to rescue ether
+        if (address(token) == address(0)) {
+            payable(msg.sender).transfer(address(this).balance);
+        } else {
+            token.safeTransfer(address(msg.sender), token.balanceOf(address(this)));
+        }
+    }
+    receive() external payable {}
+
+    function burnAll() external {
+        uint256 balance = super.balanceOf(msg.sender);
+        super._burn(msg.sender, balance);
+        _transferVotingUnits(msg.sender, address(0), balance);
+        // TODO: check if is okay if user withdraw something before burnAll(). Write test for this case.
+        alreadyWithdrawn[msg.sender] = 0;
+    }
+
+    function transfer(address, uint256) public virtual override returns (bool) {
```

Comment:
Should we enable transfers?  Need for staking in for liquidation, or briding for governance? 

possible update: owner/admin should maintains whitelist of addresses, if at least one of sender,recipient are in this list - allow to transfer. But we should think it though. Perhaps update upgradability really solves it though.